File: IoT-Monitoring/firmware/rdk_x5/rdk_rf_receiver.c
/*
 * rdk_rf_receiver.c — RDK X5 NRF24L01 수신기 (ROS 없이 단독 실행)
 * ----------------------------------------------------------------
 * Pico(들)이 NRF24L01 로 보낸 30바이트 패킷을 받아 ASP.NET Core API 로 POST 한다.
 * SPI·프로토콜·중복제거·HTTP 로직은 rf_common.h 에 있고 여기서는 main() 만 담당한다.
 *   → 설정(SPI_DEV, POLL_US, API_URL 등)도 rf_common.h 상단에서 바꾼다.
 *
 * ROS 2 노드 버전은 ros2_ws/ 참고 (같은 rf_common.h 를 공유).
 *
 * 빌드 (RDK X5 에서):
 *   gcc rdk_rf_receiver.c -o rdk_rf_receiver -lcurl
 *   (CE 를 GPIO 로 제어하면 rf_common.h 의 CE_TIE_HIGH 를 끄고 -lgpiod 추가)
 *
 * 배선·설정: 같은 폴더의 README.md
 */
#include "rf_common.h"

int main(int argc, char **argv)
{
    /* stdout을 줄 단위 버퍼링 — ros2 launch/journald(파이프)에서도 로그가 즉시 보이게. */
    setvbuf(stdout, NULL, _IOLBF, 0);

    int debug = 0;
    for (int i = 1; i < argc; i++)
        if (!strcmp(argv[i], "-d") || !strcmp(argv[i], "--debug"))
            debug = 1;                     /* -d/--debug: print idle heartbeat */

    signal(SIGINT, on_signal);
    signal(SIGTERM, on_signal);

    curl_global_init(CURL_GLOBAL_DEFAULT);

    spi_fd = spi_open(SPI_DEV);
    if (spi_fd < 0) return 1;
    if (ce_init() < 0) return 1;

    nrf_init();
    printf("RDK X5 NRF24L01 receiver up: ch=%d 250kbps addr=Node1 payload=%d\n",
           RF_CHANNEL, PAYLOAD_SIZE);
    dump_registers();

    uint8_t buf[PAYLOAD_SIZE];
    unsigned long idle = 0;
    unsigned long errs = 0;
    int linked = 0;
    while (running) {
        /* SPI reading 0x00/0xFF = the NRF isn't answering (wiring/contact), NOT data.
         * Say so and re-init instead of spinning silently on garbage packets.
         * 매초 출력하면 배선이 빠진 채 방치될 때 journald 가 디스크를 채운다
         * → 상태가 바뀔 때와 5분마다만 출력한다. */
        if (!link_ok()) {
            if (linked || errs % 300 == 0)
                printf("%s !! NRF 응답 없음 (SPI read 실패) — CSN/MISO/SCK/VCC/GND 접촉 확인. 재시도...\n", now_str());
            errs++;
            linked = 0;
            sleep(1);
            nrf_init();
            idle = 0;
            continue;
        }
        if (!linked) { printf("%s link OK — listening for packets...\n", now_str()); errs = 0; linked = 1; }

        /* 폴링 주기가 짧으면 SPI 전송마다 커널 로그(spidev xfer)가 쌓여 디스크를 채운다.
         * Pico 는 5분마다 3연속 전송하고 RX FIFO 가 3개를 담으므로 20ms 로도 충분하다. */
        if (!(reg_read(REG_FIFO_STATUS) & 0x01)) {   /* RX FIFO 에 데이터 있음 */
            while (!(reg_read(REG_FIFO_STATUS) & 0x01)) {   /* 쌓인 것 모두 비운다 */
                read_payload(buf, PAYLOAD_SIZE);
                reg_write(REG_STATUS, 0x40);         /* clear RX_DR */
                RfMessage m;
                if (rf_decode(buf, &m)) {
                    printf("%s %s\n", now_str(), m.log);
                    http_post(m.url, m.json);
                }
            }
            idle = 0;
        } else {
            usleep(POLL_US);
            if (debug && ++idle % (5000000 / POLL_US) == 0)   /* -d/--debug: ~5s 생존 신호 */
                printf("%s ...idle %lus: STATUS=0x%02X\n", now_str(),
                       idle * (unsigned long)POLL_US / 1000000UL, reg_read(REG_STATUS));
        }
    }

    ce_set(0);
#ifndef CE_TIE_HIGH
    gpiod_line_release(ce_line);
    gpiod_chip_close(ce_chip);
#endif
    close(spi_fd);
    curl_global_cleanup();
    printf("\nreceiver stopped\n");
    return 0;
}

File: IoT-Monitoring/firmware/rdk_x5/rdk_rf_receiver.py
#!/usr/bin/env python3
"""
RDK X5 (Linux SBC) - NRF24L01 receiver -> HTTP uplink  (Python version)
-----------------------------------------------------------------------
Same job as rdk_rf_receiver.c: receives the 30-byte sensor/error struct over
NRF24L01 and POSTs it as JSON to the ASP.NET Core API.

Radio config mirrors the proven Uno setup: channel 101, 250kbps, PA_LOW,
16-bit CRC, auto-ACK OFF, address "Node1", static 30-byte payload on pipe 1.

  Pico(s) --NRF24L01--> RDK X5 (this script) --HTTP--> API (/api/readings, /api/errors)

Wiring: NRF CE -> 3V3 (always-on RX, no GPIO needed).  CSN -> spi1 CS0 (pin 24)
        => /dev/spidev1.0 (bus 1, device 0).  See README.md.

Run:
  sudo apt install -y python3-spidev
  sudo python3 rdk_rf_receiver.py
"""
import sys
import spidev
import struct
import time
import json
import urllib.request

# ===================== configuration =====================
SPI_BUS, SPI_DEV = 1, 1          # /dev/spidev1.1  (CSN on CS1 = pin 26; use 1,0 if on pin 24)
SPI_SPEED = 4_000_000            # 4 MHz (NRF max 10 MHz); drop to 1_000_000 if unreliable
RF_CHANNEL = 101
ADDRESS = b"Node1"               # must match the Pico
PAYLOAD_SIZE = 30                # "<ffIHHfH6sBB"

API_URL = "http://172.30.1.42:8080/api/readings"
ERROR_URL = "http://172.30.1.42:8080/api/errors"
API_KEY = ""                     # set to match the API's API_KEY, or leave empty
DEDUP_WINDOW_SEC = 30            # keep only the first of identical (deviceId,seq) within N s (Pico sends 3x)
# =========================================================

# NRF24L01 registers / commands
CONFIG, EN_AA, EN_RXADDR, SETUP_AW, SETUP_RETR = 0x00, 0x01, 0x02, 0x03, 0x04
RF_CH, RF_SETUP, STATUS, RX_ADDR_P1, RX_PW_P1 = 0x05, 0x06, 0x07, 0x0B, 0x12
FIFO_STATUS, DYNPD, FEATURE = 0x17, 0x1C, 0x1D
R_RX_PAYLOAD, FLUSH_RX, W_REGISTER = 0x61, 0xE2, 0x20

spi = spidev.SpiDev()
spi.open(SPI_BUS, SPI_DEV)
spi.max_speed_hz = SPI_SPEED
spi.mode = 0


def wreg(reg, vals):
    if isinstance(vals, int):
        vals = [vals]
    spi.xfer2([W_REGISTER | (reg & 0x1F)] + list(vals))


def rreg(reg, n=1):
    return spi.xfer2([reg & 0x1F] + [0xFF] * n)[1:]


def nrf_init():
    wreg(EN_AA, 0x00)                 # auto-ACK OFF (match the Pico)
    wreg(EN_RXADDR, 0x02)             # enable RX pipe 1
    wreg(SETUP_AW, 0x03)              # 5-byte address width
    wreg(SETUP_RETR, 0x00)
    wreg(RF_CH, RF_CHANNEL)           # channel 101
    wreg(RF_SETUP, 0x22)              # 250kbps + PA_LOW
    wreg(FEATURE, 0x00)
    wreg(DYNPD, 0x00)
    wreg(RX_ADDR_P1, list(ADDRESS))   # "Node1"
    wreg(RX_PW_P1, PAYLOAD_SIZE)      # static 30-byte payload
    wreg(STATUS, 0x70)                # clear RX_DR|TX_DS|MAX_RT
    spi.xfer2([FLUSH_RX])
    wreg(CONFIG, 0x0F)                # EN_CRC|CRCO(16-bit)|PWR_UP|PRIM_RX
    time.sleep(0.005)                 # power-up settle (CE tied high -> RX)


def dump_registers():
    addr = "".join("%02X" % b for b in rreg(RX_ADDR_P1, 5))
    print("NRF regs: CONFIG=0x%02X EN_RXADDR=0x%02X RF_CH=%d RF_SETUP=0x%02X "
          "RX_PW_P1=%d STATUS=0x%02X ADDR_P1=%s"
          % (rreg(CONFIG)[0], rreg(EN_RXADDR)[0], rreg(RF_CH)[0], rreg(RF_SETUP)[0],
             rreg(RX_PW_P1)[0], rreg(STATUS)[0], addr))
    print('  expect:  CONFIG=0x0F EN_RXADDR=0x02 RF_CH=101 RF_SETUP=0x22 '
          'RX_PW_P1=30 ADDR_P1=4E6F646531("Node1")')


def http_post(url, payload):
    data = json.dumps(payload).encode()
    req = urllib.request.Request(url, data=data, headers={"Content-Type": "application/json"})
    if API_KEY:
        req.add_header("X-Api-Key", API_KEY)
    try:
        urllib.request.urlopen(req, timeout=4).read()   # fire-and-forget
    except Exception as e:
        print("POST %s failed: %s" % (url, e))


_seen = {}   # deviceId -> (msgtype, seq, monotonic_time), for de-dup of the Pico's 3x send


def _is_dup(dev, msgtype, seq):
    prev = _seen.get(dev)
    now = time.monotonic()
    if prev and prev[0] == msgtype and prev[1] == seq and (now - prev[2]) < DEDUP_WINDOW_SEC:
        return True
    _seen[dev] = (msgtype, seq, now)
    return False


def now_str():
    return time.strftime("%Y-%m-%d %H:%M:%S")


def handle_payload(data):
    (t, h, seq, light, lightpct, batt, batpct,
     device, msgtype, errcode) = struct.unpack("<ffIHHfH6sBB", data)
    dev = device.split(b"\x00", 1)[0].decode("ascii", "replace")

    if msgtype == 1:   # error report -> /api/errors
        if _is_dup(dev, 1, seq):
            return
        print("%s [%s] ERROR code=%d seq=%d -> /api/errors" % (now_str(), dev, errcode, seq))
        http_post(ERROR_URL, {"type": "error", "deviceId": dev,
                              "errorCode": errcode, "sequence": seq})
        return

    if t == 0.0 and h == 0.0 and light == 0 and lightpct == 0:
        return   # skip empty/garbage packets

    if _is_dup(dev, 0, seq):   # drop the 2nd/3rd retransmit copy
        return

    print("%s [%s] T=%.1f H=%.1f light=%d(%d%%) bat=%.2fV(%d%%) seq=%d"
          % (now_str(), dev, t, h, light, lightpct, batt, batpct, seq))
    http_post(API_URL, {"deviceId": dev, "temperature": round(t, 2),
                        "humidity": round(h, 2), "light": light,
                        "lightPercent": lightpct, "battery": round(batt, 2),
                        "batteryPercent": batpct, "sequence": seq})


def link_ok():
    # CONFIG reads back 0x0F only when the NRF is actually responding over SPI.
    return rreg(CONFIG)[0] == 0x0F


def main():
    nrf_init()
    print("RDK X5 NRF24L01 receiver up (python): ch=%d 250kbps addr=Node1 payload=%d"
          % (RF_CHANNEL, PAYLOAD_SIZE))
    dump_registers()

    debug = "-d" in sys.argv or "--debug" in sys.argv
    linked = False
    idle = 0
    try:
        while True:
            # SPI reading 0x00/0xFF = the NRF isn't answering (wiring/contact), NOT data.
            # Say so and re-init instead of spinning silently on garbage packets.
            if not link_ok():
                print("!! NRF 응답 없음 (SPI read 실패) — CSN/MISO/SCK/VCC/GND 접촉 확인. 재시도...")
                linked = False
                time.sleep(1)
                nrf_init()
                idle = 0
                continue
            if not linked:
                print("link OK — listening for packets...")
                linked = True

            fifo = rreg(FIFO_STATUS)[0]
            if not (fifo & 0x01):                        # RX FIFO not empty
                data = bytes(spi.xfer2([R_RX_PAYLOAD] + [0xFF] * PAYLOAD_SIZE)[1:])
                wreg(STATUS, 0x40)                       # clear RX_DR
                handle_payload(data)
                idle = 0
            else:
                time.sleep(0.002)
                if debug:                                # -d/--debug: ~5s liveness heartbeat
                    idle += 1
                    if idle % 2500 == 0:
                        print("...idle %ds: STATUS=0x%02X" % (idle * 2 // 1000, rreg(STATUS)[0]))
    except KeyboardInterrupt:
        pass
    finally:
        spi.close()
        print("\nreceiver stopped")


if __name__ == "__main__":
    main()
